integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,49 @@
+package kotlinx.coroutines.androidx.lifecycle
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.State.INITIALIZED
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.*
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that will be cancelled as
+ * soon as this [Lifecycle] [currentState][Lifecycle.getCurrentState] is no longer
+ * [at least][Lifecycle.State.isAtLeast] the passed [activeWhile] state.
+ *
+ * **Beware**: if the current state is lower than the passed [activeWhile] state, you'll get an
```

Comment:
This API surprises me in two ways:

1) If I instantiate this very early (e.g. in field initialization) it will always end up creating a cancelled job which I won't discover until my coroutine doesn't run.
2) If a state is re-enterable (e.g. resumed on Fragment) this will not resume again.

What do you think a lifecycle aware dispatcher to solve both of those cases as well as handle the onCreate situation discussed previously?

I'm not sure this exactly the right API, but it seems likely to work out for these cases. Basically, it'd operate as an event loop that pauses when it's below the expected state. What do you think?

Reply:
If you want to launch a coroutine at instantiation time with a lifecycle aware scope, you can still use `createScope(activeWhile = Lifecycle.State.INITIALIZED)`.

Do you think I should change the default for the `coroutineScope` and `job` properties to be active while initialized instead of active while created?

Reply:
About your second point: the typical use case there is to launch a coroutine inside the `onStart` or `onResume` method that is automatically cancelled when the lifecycle is `paused` or `resumed`, so a new coroutine is launched each time the `onStart` or `onResume` method is called.

Reply:
When a `Job` is cancelled, there's no going back, but you can always start a coroutine later when the conditions are met again, as mentioned in my comment just above.

About pausable dispatcher, I thought about this a while ago, but this doesn't fit well with coroutines design.

You could think that you could pause dispatching, but this could cause more problems, as you may be relying on a dispatching loop to release resources, creating temporary leaks in suspended coroutines that may resume only much later, if ever.

Suspending work based on the lifecycle has to be opt-in at call site.
This is something that is possible with [this `awaitState(…)` extension function for `Lifecycle` I wrote](https://github.com/LouisCAD/Splitties/blob/54261e937177224dca1a5a9f81690eb48d3be624/sample/src/main/java/com/louiscad/splittiessample/extensions/coroutines/LifecycleAwaitState.kt#L23), and that can work well for the use cases where you want to pause execution until the lifecycle is at least in some state. You can call this before starting expensive work or in a loop tied to the UI component (Activity, Fragment…).

I personally already have use cases for this method, like[ showing `DialogFragment`s safely](https://github.com/LouisCAD/Splitties/blob/54261e937177224dca1a5a9f81690eb48d3be624/sample/src/main/java/com/louiscad/splittiessample/extensions/coroutines/ShowDialogFragments.kt#L26-L72).
Do you think I should include this method in this PR?